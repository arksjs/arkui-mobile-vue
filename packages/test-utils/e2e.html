<script src="../../node_modules/vue/dist/vue.global.js"></script>
<script src="../../dist/index.js"></script>
<link rel="stylesheet" href="../../dist/index.css" />

<div id="app"></div>
