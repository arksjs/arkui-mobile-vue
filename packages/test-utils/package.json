{
  "name": "@arksjs/test-utils",
  "version": "0.0.1",
  "description": "arkui test-utils",
  "main": "dist/index.js",
  "private": true
}
