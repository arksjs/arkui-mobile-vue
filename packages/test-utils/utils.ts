export const timeout = (n: number) => new Promise(r => setTimeout(r, n))
