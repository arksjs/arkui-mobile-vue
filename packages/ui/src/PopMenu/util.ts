import { isObject } from '../helpers/util'
import type { Option } from './types'

export const getOptions = (options?: Option[]) => {
  const newOptions: Option[] = []

  if (Array.isArray(options)) {
    options.forEach(v => {
      newOptions.push(
        isObject(v)
          ? {
              icon: v.icon || undefined,
              name: v.name,
              disabled: !!v.disabled
            }
          : {
              icon: undefined,
              name: v.toString(),
              disabled: false
            }
      )
    })
  }

  return newOptions
}

export const getItemClasses = (option: Option) => {
  return [
    'ak-pop-menu_item',
    'ak-horizontal-hairline',
    { disabled: !!option.disabled, 'has--icon': !!option.icon }
  ]
}
