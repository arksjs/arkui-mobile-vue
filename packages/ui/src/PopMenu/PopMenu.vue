<template>
  <Popover
    class="ta-pop-dialog"
    :visible="visible"
    :selector="selector"
    :placement="placement"
    :showMask="showMask"
    @visibleStateChange="onVisibleStateChange"
    @confirm="onConfirm"
    @cancel="onCancel"
    @update:visible="onUpdateVisible"
    ref="popupRef"
  >
    <ul
      class="ta-pop-menu_list ta-pop-menu_list"
      v-if="options2 && options2.length > 0"
    >
      <li
        :class="getItemClasses(item)"
        v-for="(item, index) in options2"
        :key="index"
        @click="onItemClick(index)"
      >
        <div class="ta-pop-menu_item-inner">
          <Icon v-if="item.icon" :icon="item.icon" />
          <span>{{ item.name }}</span>
        </div>
      </li>
    </ul>
  </Popover>
</template>

<script lang="ts">
import { computed, defineComponent, type PropType } from 'vue'
import { Popover } from '../Popover'
import { Icon } from '../Icon'
import type { Option, Detail, PopMenuEmits } from './types'
import { getOptions, getItemClasses } from './util'
import { isNumber, isString, type PropsToEmits } from '../helpers'
import { popoverProps, popoverEmits } from '../Popover/props'
import { usePopupExtend } from '../popup/use-popup'

export default defineComponent({
  name: 'ta-pop-menu',
  components: { Icon, Popover },
  props: {
    ...popoverProps,
    options: {
      type: Array as PropType<Option[]>
    }
  },
  emits: {
    ...popoverEmits,
    confirm: payload =>
      payload &&
      isNumber(payload.index) &&
      payload.item &&
      isString(payload.item.name)
  } as PropsToEmits<PopMenuEmits>,
  setup(props, ctx) {
    const popup = usePopupExtend<Detail>(ctx)

    function onItemClick(index: number) {
      const item = props.options?.[index]

      if (!item || item.disabled) {
        return
      }

      popup.customConfirm({
        item: {
          name: item.name
        },
        index
      })
    }

    const options2 = computed(() => getOptions(props.options))

    return {
      ...popup,
      onItemClick,
      options2,
      getItemClasses
    }
  }
})
</script>
