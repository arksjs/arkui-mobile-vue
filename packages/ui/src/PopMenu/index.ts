import { withInstall } from '../helpers/with-install'
import PopMenu from './PopMenu.vue'
import { createConfirmHook, createShowPopup } from '../popup/api'
import type { PopupSuccessConfirmArgs } from '../popup/types'
import type { Selector, PlacementType } from '../helpers/types'
import type { Option, Detail } from './types'

const showPopMenu = createShowPopup<
  {
    selector: Selector
    options: Option[]
    placement?: PlacementType
  },
  PopupSuccessConfirmArgs<Detail>
>({
  apiName: 'showPopMenu',
  component: PopMenu,
  createHook: createConfirmHook
})

export { PopMenu, showPopMenu }
export default PopMenu
export const install = withInstall(PopMenu)
