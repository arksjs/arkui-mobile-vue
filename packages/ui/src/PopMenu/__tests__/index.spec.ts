import { describe, test, expect } from 'vitest'
import { mount, shallowMount } from '@vue/test-utils'
import { PopMenu } from '@/PopMenu'

const options = [
  {
    icon: 'HeartOutlined',
    name: 'Heart'
  },
  {
    icon: 'StarOutlined',
    name: 'Star'
  },
  {
    icon: 'CircleOutlined',
    name: 'Circle',
    disabled: true
  }
]

const selector = 'body'

describe('PopMenu', () => {
  test('snapshot', () => {
    const wrapper = shallowMount(PopMenu, {
      global: {
        stubs: { teleport: true }
      },
      props: {
        selector
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  test('set options', () => {
    const wrapper = mount(PopMenu, {
      global: {
        stubs: { teleport: true }
      },
      props: {
        options,
        selector
      }
    })

    expect(wrapper.findAll('.ta-pop-menu_item').length).toBe(3)
    expect(wrapper.findAll('.ta-pop-menu_item')[0].text()).toBe('Heart')
    expect(wrapper.findAll('.ta-pop-menu_item')[1].text()).toBe('Star')
    expect(wrapper.findAll('.ta-pop-menu_item')[2].text()).toBe('Circle')
  })

  test.concurrent('showMask', async () => {
    const wrapper = mount(PopMenu, {
      global: {
        stubs: { teleport: true }
      },
      props: {
        selector
      }
    })
    expect(wrapper.find('.ta-popover').classes('no--mask')).toBeFalsy()

    await wrapper.setProps({
      showMask: false
    })
    expect(wrapper.find('.ta-popover').classes('no--mask')).toBeTruthy()
  })
})
