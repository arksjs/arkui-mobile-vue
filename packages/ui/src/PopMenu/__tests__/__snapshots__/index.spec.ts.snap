// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PopMenu > snapshot 1`] = `"<popover-stub selector=\\"body\\" visible=\\"false\\" maskclosable=\\"true\\" placement=\\"bottom\\" showmask=\\"true\\" content=\\"\\" class=\\"ta-pop-dialog\\"></popover-stub>"`;
